### DISCUSSION TITLE
Using Next.js for server-side generation of share preview

### DISCUSSION THREAD

**Discussion by mpeterdev** [NearSocial/viewer#112] [General]
If the Near Social Viewer project is converted to Next.js, we could use [dynamic routes](https://nextjs.org/docs/routing/dynamic-routes) to fetch component metadata and use that metadata in social share previews.

~I do not believe this could be used to give granular previews to specific posts, but info about the #app is a step in the right direction~

**Comment by mpeterdev** [comment]
@calebjacob can you take a quick look into whether query params—which are used to direct to individual posts in a Near Social app—might be leveraged on-demand for generating share previews?

**Reply by calebjacob** [reply]
Sure thing! I'll run some tests. I would guess the answer is yes.

**Reply by calebjacob** [reply]
@mpeterdev Yes, we can use both URL segment params (`/:slug/...`) and query params (`?foo=bar`) inside Next's `getServerSideProps()` method.

I threw together an example Next app that fetches post data from NEAR Social via a view call to dynamically render `<meta>` tags via SSR: https://near-next-dynamic-meta-tags.vercel.app/hello?accountId=root.near&blockHeight=83395215

[NEAR Social URL](https://near.social/#/mob.near/widget/MainPage.Post.Page?accountId=root.near&blockHeight=83395215)

Here's the corresponding code for this page: https://github.com/calebjacob/near-next-dynamic-meta-tags/blob/main/src/pages/%5Bslug%5D.tsx

_NOTE: The image URL is hardcoded just for demo purposes but could easily be dynamic if a NEAR view call provided a full image URL._

<img width="722" alt="Screen Shot 2023-01-23 at 3 03 03 PM" src="https://user-images.githubusercontent.com/1475067/214160004-3ead1060-8581-48a9-aba0-5db9c1a94e5a.png">

**Reply by mpeterdev** [reply]
awesome 🔥 

I think the major remaining question would be how a widget specifies the content to be pulled from SocialDB based on their data schema, right?

**Reply by calebjacob** [reply]
That raises an interesting question. Should we expose the direct use of `getServerSideProps()` to widget developers? We might need to consider letting users choose if their widget is a Next JS page component or just a regular component. This would probably introduce new attack vectors to consider (leaking environment variables).

The alternative is to expose some type of abstraction for `getServerSideProps()`, but then you start heading further down the road of non-standard API's.

**Reply by mpeterdev** [reply]
Initial impression is that exposing an abstraction would be the way to go, since otherwise we give component developers the ability to run arbitrary JS in the server environment which should be considered secure

**Reply by evgenykuzyakov** [reply]
What's the abstraction level there? To fetch the post you may need to fetch data from chain multiple times, e.g. fetch the author profile and the post itself. The ideal solution is to have some smaller JS code that you would run on the server side, but within some VM (maybe tiny JS and wasm). Then limit this VM by memory and compute time.

Short term we can use app metadata to at least display the name of the app, description and icon, but not the content being shared.

**Reply by calebjacob** [reply]
That would definitely be a good short term win @evgenykuzyakov.
